docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,331 @@
+- A TPM provider MUST annotate new CSRs as follows:
+    - Key: cluster-x-k8s-io/tpm-endorsement-key
+    - Value: Platform-specific endorsement key (e.g., retrieved from GCP Shielded VM API or VMware vCenter)
+
+#### Kubeadm
+Since this proposal essentially takes over part of the node registration process from kubeadm, we will require the following changes:
+- kubeadm COULD allow opt-out of kubeadm setting up ClusterRoleBindings between the system:nodes group and the `system:certificates.k8s.io:certificatesigningrequests:selfnodeclient` permission, so that certificate renewals must go through re-attestation.
+- Kubeadm COULD allow opt-out of kubeadm setting up the bootstrapper and node groups and bindings.
+
+This is deferred to a future Kubeadm design and release, and for this proposal, we will add fields to KubeadmControlPlane to remove these node groups and bindings post control plane initialisation.
+
+#### Changes to KubeadmControlPlane
+
+```
+
+```
+
+#### Kubelet authenticator flow
+
+The authenticator will be responsible for both updating the kubelet client and serving certificates.
+
+Both of these decision flows will operate inside the authenticator, and execute whenever called by kubelet to provide client authentication.
+
+##### Client authentication flow
+
+![client auth](images/kubelet-authentication/client-authenticator-flow.png)
+
+##### Serving certificate flow
+
+![serving cert flow](images/kubelet-authentication/serving-certificate-flow.png)
+
+### User Stories
+
+#### Story 1: machine attestation
+
+A cluster operator has been asked to ensure compliance with [NIST SP 800-190 Application Container Security][nist-sp-800-190] Guide. Hardware countermeasure 4.6 suggests that container platforms should make use of trusted computing. In a Kubernetes context, this would mean providing hardware node attestation wherever possible.
+
+### Requirements (Optional)
+
+The node bootstrapper MUST be able to attest the identity of the machine against a chain of trust provided by the hardware or cloud provider.
+
+### Implementation Details/Notes/Constraints
+
+
+### Security Model
+
+Document the intended security model for the proposal, including implications
+on the Kubernetes RBAC model. Questions you may want to answer include:
+
+* Does this proposal implement security controls or require the need to do so?
+  * If so, consider describing the different roles and permissions with tables.
+* Are there adequate security warnings where appropriate (see https://adam.shostack.org/ReederEtAl_NEATatMicrosoft.pdf for guidance)?
+* Are regex expressions going to be used, and are their appropriate defenses against DOS?
+* Is any sensitive data being stored in a secret, and only exists for as long as necessary?
+
+### Risks and Mitigations
+
+There may be additional security risks being introduced in this design. In order to mitigate this, this proposal will be taken to SIG Security and SIG Auth for review **before the beta graduation**.
```

Comment:
i don't see anyone from SIG Auth or SIG Security in the reviewers (might have missed it)
@randomvariable @yastij do you think it's a good idea to have such a third party review?
once this is added as alpha, i can see users starting to use it immediately, in production.

Reply:
@neolit123 @vincepri - ideally we should add a featuregate for this; I think we can have a proper review from auth/security before going to beta.

Reply:
@neolit123 @vincepri - ideally we should add a featuregate for this; I think we can have a proper review from auth/security before turning the feature-gate to be always on, thoughts ?

Reply:
We should probably have SIG Auth or Security review it even if it's in alpha, @dims @neolit123 could you help ping the appropriate folks?

Reply:
an FG makes sense to me.

Reply:
we should start with @enj our resident guru :)

Reply:
+1 for @enj 

it would be nice to get a non-VMware review as well.
i will add a TODO to ask next week.

Reply:
i've send pings to the sig-auth and sig-security slack channels.

Reply:
@vincepri @yastij are we waiting for sig-auth / sig-security reviews before merging the proposal? If so it might delay things quite a bit. Not a problem but we should proceed with v0.4.0 and make sure this is non-blocking.